src/Parser/Core/Modules/Spells/BFA/AzeriteTraits/SecondaryStatProcTrait.js

```diff
@@ -0,0 +1,143 @@
+    statTracker: StatTracker,
+  };
+
+  // info about trait passed in by subclass
+  traitId;
+  traitName;
+  traitIcon;
+  buffId;
+  stats;
+
+  // store trait ranks from combatant
+  traitRanks;
+
+  constructor(traitId, traitName, traitIcon, buffId, stats, ...args) {
+    super(...args);
+
+    this.traitId = traitId;
+    this.traitName = traitName;
+    this.traitIcon = traitIcon;
+    this.buffId = buffId;
+    this.stats = stats;
+
+    this.active = this.selectedCombatant.hasTrait(this.traitId);
+    if (!this.active) {
+      return;
+    }
+
+    this.traitRanks = this.selectedCombatant.traitRanks(this.traitId);
+
+    // sanity check to double check the subclass is implemented correctly matching the data we have about the trait
+    if (this.stats.length !== this.statBuffs.length) {
+      throw new Error("Mismatch between stat buffs specified in azerite trait effects and passed in from subclass");
+    }
+  }
+
+  get statBuffs() {
+    return Object.values(this.traitRanks).reduce((stats, rank) => {
+      calculateAzeriteEffects(this.traitId, rank).forEach((statBuff, idx) => {
+        stats[idx] += statBuff;
+      });
+
+      return stats;
+    }, calculateAzeriteEffects(this.traitId, 0).map(() => 0));
+  }
+
+  get uptime() {
+    return this.selectedCombatant.getBuffUptime(this.buffId) / this.owner.fightDuration;
+  }
+
+  /**
+   * to make the tooltip reflect multiple ranks of the trait we use a bit of a hacky way to figure out which ilvl would result in displaying the sum of the bonus damage
+   */
+  determineIlvlForSum() {
+    let ilvl = 0;
+    if (this.traitRanks.length === 1) {
+      ilvl = this.traitRanks[0];
+    } else {
+      const desiredBonus = this.traitRanks.map((rank) => calculateAzeriteEffects(this.traitId, rank)[0]).reduce((total, bonus) => total + bonus, 0);
+      let ilvlDmgBonus = calculateAzeriteEffects(this.traitId, this.traitRanks[0])[0];
+      for (ilvl = this.traitRanks[0]; ilvlDmgBonus < desiredBonus; ilvl++) {
+        ilvlDmgBonus = calculateAzeriteEffects(this.traitId, ilvl)[0];
+      }
+    }
+
+    return ilvl;
+  }
+
+  statistic() {
+    const statBuffs = this.statBuffs;
+
+    const statStrings = this.stats.map((stat, idx) => {
+      const statBuff = statBuffs[idx];
+      const avgStat = this.uptime * statBuff;
+      const statInfo = STATS[stat];
+      const percentageFn = statInfo.percentageFn;
+      const avgPercentage = this.statTracker[percentageFn](avgStat);
+
+      return [
+        `${formatPercentage(avgPercentage)}% avg ${statInfo.name}`,
+        `<b>${statBuff} ${statInfo.name} Rating</b>`,
+        `<b>${formatNumber(avgStat)}</b> average Haste Rating (${formatPercentage(avgPercentage)}%)`,
+      ];
+    });
+
+    const valueStrings = statStrings.map((strings) => strings[0]).join(' and ');
+    const tooltip1Strings = statStrings.map((strings) => strings[1]).join(' and ');
+    const tooltip2Strings = statStrings.map((strings) => strings[2]).join(' and ');
```

Comment:
above code is a bit ... idk if someone can think of something cleaner? I'm not much of a map reduce flatten ninja.